wenet/whisper/whisper.py

```diff
@@ -0,0 +1,107 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Modified from [Whisper](https://github.com/openai/whisper)
+
+import torch
+
+from typing import Tuple
+from whisper.tokenizer import get_tokenizer
+
+from wenet.transformer.asr_model import ASRModel
+from wenet.transformer.ctc import CTC
+from wenet.transformer.encoder import TransformerEncoder
+from wenet.transformer.decoder import TransformerDecoder
+from wenet.utils.common import (IGNORE_ID, add_whisper_tokens,
+                                reverse_pad_list, th_accuracy)
+
+
+class Whisper(ASRModel):
+    def __init__(
+        self,
+        vocab_size: int,
+        encoder: TransformerEncoder,
+        decoder: TransformerDecoder,
+        ctc: CTC = None,
+        ctc_weight: float = 0.5,
+        ignore_id: int = IGNORE_ID,
+        reverse_weight: float = 0.0,
+        lsm_weight: float = 0.0,
+        length_normalized_loss: bool = False,
+    ):
+        super().__init__(vocab_size, encoder, decoder, ctc, ctc_weight, ignore_id,
+                         reverse_weight, lsm_weight, length_normalized_loss)
+        self.tokenizer = get_tokenizer(multilingual=self.is_multilingual,
+                                       num_languages=self.num_languages)
+        assert vocab_size == self.tokenizer.encoding.n_vocab, "{} v.s. {}".format(
+            vocab_size, self.tokenizer.encoding.n_vocab)
+        self.sos = self.tokenizer.sot
+        self.eos = self.tokenizer.eot
+
+    # TODO(xcsong): time align
+    def set_alignment_heads(self, dump: bytes):
+        raise NotImplementedError
+
+    @property
+    def device(self):
+        return next(self.parameters()).device
+
+    @property
+    def is_multilingual(self):
+        return self.vocab_size >= 51865
+
+    @property
+    def num_languages(self):
+        return self.vocab_size - 51765 - int(self.is_multilingual)
+
+    def _calc_att_loss(
+        self,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+        ys_pad: torch.Tensor,
+        ys_pad_lens: torch.Tensor,
+    ) -> Tuple[torch.Tensor, float]:
+        # TODO(xcsong): add args for no_timestamp, language, etc
+        prev_len = ys_pad.size(1)
+        ys_in_pad, ys_out_pad = add_whisper_tokens(
+            self.tokenizer, ys_pad, self.ignore_id, task_id=self.tokenizer.transcribe,
+            no_timestamp=True, language="zh", use_prev=False
+        )
+        cur_len = ys_in_pad.size(1)
+        ys_in_lens = ys_pad_lens + cur_len - prev_len
+
+        # reverse the seq, used for right to left decoder
```

Comment:
there  is no right-to-left decoder in whisper? should we drop the right-to-left logic here to keep the code as less as possible?

Reply:
okay !